editor/src/clj/util/http_server.clj

```diff
@@ -13,98 +13,415 @@
+                              connection (data-connection body)
+                              ;; Infer content-length and content-type for responses that
+                              ;; can't do that during response creation. For example, we
+                              ;; might cache a response with a particular file, but
+                              ;; the file itself might change between requests
+                              headers (-> headers
+                                          (provide-header "content-type" (content-type connection))
+                                          (provide-header "content-length" (data-length connection)))
+
+                              connection (if (or (= "HEAD" request-method)
+                                                 (= 304 status))
+                                           (do
+                                             (when (instance? AutoCloseable connection)
+                                               (.close ^AutoCloseable connection))
+                                             nil)
+                                           connection)]
+                          (when-not (integer? status)
+                            (throw (ex-info (str "Invalid response status: " status) {:status status})))
+                          (when-not (or (nil? headers) (map? headers))
+                            (throw (ex-info (str "Invalid response headers: " headers) {:headers headers})))
+                          (reduce-kv #(doto ^Headers %1 (.add %2 %3)) (.getResponseHeaders exchange) headers)
+                          (.sendResponseHeaders
+                            exchange
+                            status
+                            ;; Response length argument:
+                            ;; -1 means empty response body (0 bytes)
+                            ;; 0 is unknown size (chunked)
+                            ;; positive is exact size
+                            (if (nil? connection)
+                              -1
+                              (if-let [explicit-length (get headers "content-length")]
+                                (let [n (Long/parseUnsignedLong explicit-length)]
+                                  (if (zero? n) -1 n))
+                                0)))
+                          (when connection
+                            (try
+                              (with-open [is (io/input-stream connection)]
+                                (try
+                                  (io/copy is (.getResponseBody exchange))
+                                  ;; A browser or remote http client can close
+                                  ;; the connection before we finished sending
+                                  ;; the response: ignore such exceptions since
+                                  ;; it's not an issue.
+                                  (catch IOException _)))
```

Comment:
According to the docs for `getResponseBody`:

> If too many bytes are written, then the write method of `OutputStream` will throw an `IOException`. If too few bytes are written then the stream `OutputStream.close()` will throw an `IOException`.

I think the case you're catching here is when _too many_ bytes are written to the output stream returned by `.getResponseBody`? `io/copy` doesn't close the stream, so presumably that won't happen until `.close` is called on the `exchange` in the `finally` clause below?

Reply:
Hmm, that's a good point. The actual exceptions I was getting were `Broken pipe` and `Connection reset by peer` when I was holding `F5` in the browser, which caused it to constantly reload the engine profiler page. I think I'll log the exceptions instead of ignoring them completely (in another PR)